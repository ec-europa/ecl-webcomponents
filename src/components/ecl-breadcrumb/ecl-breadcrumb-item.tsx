import { Component, h, Prop} from '@stencil/core';

@Component({
  tag: 'ecl-breadcrumb-item',
  shadow: false,
  scoped: true,
})
export class EclBreadcrumbItem {
  @Prop() theme: string = 'ec';
  @Prop() styleClass: string;
  @Prop() variant: string;
  @Prop() path: string;
  @Prop() currentPage: boolean = false;
  @Prop() ellipsis: boolean = false;
  @Prop() buttonAriaLabel: string = '';

  getClass(): string {
    const styleClasses = [
      `ecl-breadcrumb__segment`,
      `sc-ecl-breadcrumb-${this.theme}`,
      this.styleClass
    ];

    if (this.ellipsis) {
      styleClasses.push('ecl-breadcrumb__segment--ellipsis');
    }

    return styleClasses.join(' ');
  }

  getLinkClass(): string {
    const linkClasses = [
      `sc-ecl-breadcrumb-${this.theme}`,
      'ecl-breadcrumb__link',
      'ecl-link--no-visited',
    ];

    if (this.variant === 'negative') {
      linkClasses.push('ecl-link--negative');
    }

    return linkClasses.join(' ');
  }

  getLiAttrs() {
    const attrs = { 'data-ecl-breadcrumb-item' : 'static' };
    if (this.ellipsis) {
      attrs['data-ecl-breadcrumb-ellipsis'] = 'data-ecl-breadcrumb-ellipsis';
    }

    return attrs;
  }

  render() {
    return (
      <li
        class={this.getClass()}
        {...this.getLiAttrs()}
      >
      { !this.currentPage && !this.ellipsis ? 
        <ecl-link
          variant="standalone"
          path={this.path}
          style-class={this.getLinkClass()}
        >
          <ecl-icon
            slot="icon-after"
            style-class={`ecl-breadcrumb__icon sc-ecl-breadcrumb-${this.theme}`}
            icon="corner-arrow"
            transform="rotate-90"
            size="2xs"
          ></ecl-icon>
          <slot></slot>
        </ecl-link> :
        <slot></slot>
      }
      { this.ellipsis ? 
        <ecl-button
          style-class={`ecl-breadcrumb__ellipsis sc-ecl-breadcrumb-${this.theme}`}
          variant="ghost"
          data-ecl-breadcrumb-ellipsis-button
          aria-label={this.buttonAriaLabel}
        >
          ...
          <ecl-icon
            style-class={`ecl-breadcrumb__icon sc-ecl-breadcrumb-${this.theme}`}
            size="2xs"
            transform="rotate-90"
            icon="corner-arrow"
            slot="icon-after"
          ></ecl-icon>
        </ecl-button> : '' 
      }
      { this.currentPage ?
        <slot></slot> : ''
      }
      </li>
    );
  }
}