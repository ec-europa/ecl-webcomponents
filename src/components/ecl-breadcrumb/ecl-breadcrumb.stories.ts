const getArgs = () => {
  return {
    variant: 'negative',
  };
};

const getArgTypes = () => {
  return {
    variant: {
      type: { name: 'select' },
      options: ['default', 'negative'],
      description: "Breadcrumb variant",
    },
  };
};

export default {
  title: 'Components/Navigation/breadcrumb',
};

const Template = (args) => 
  `<ecl-breadcrumb
    variant="${args.variant}"
    theme="${args.theme}"
    ecl-script
   >
    <ecl-breadcrumb-item
      path="/example.html"
      theme="${args.theme}"
    >
      Home
    </ecl-breadcrumb-item>
    <ecl-breadcrumb-item
      ellipsis
      button-aria-label="Click to expand"
     >
     </ecl-breadcrumb-item>
    <ecl-breadcrumb-item
      path="/example.html"
      theme="${args.theme}"
    >
      About the European Commission
    </ecl-breadcrumb-item>
    <ecl-breadcrumb-item
      path="/example.html"
      theme="${args.theme}"
    >
      Organisational structure
    </ecl-breadcrumb-item>
    <ecl-breadcrumb-item
      path="/example.html"
      theme="${args.theme}"
    >
      How the Commission is organised
    </ecl-breadcrumb-item>
    <ecl-breadcrumb-item
      current-page
      theme="${args.theme}"
    >
      News
    </ecl-breadcrumb-item>
  </ecl-breadcrumb>`;

export const Breadcrumb = Template.bind({});
Breadcrumb.storyName = 'default';
Breadcrumb.args = getArgs();
Breadcrumb.argTypes = getArgTypes();