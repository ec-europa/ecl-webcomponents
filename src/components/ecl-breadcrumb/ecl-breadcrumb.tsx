import { Component, Prop, h, Element } from '@stencil/core';
import getAssetPath from "../../utils/assetPath";
declare const ECL: any;

@Component({
  tag: 'ecl-breadcrumb',
  styleUrls: {
    ec: './build/styles/ecl-breadcrumb-ec.css',
    eu: './build/styles/ecl-breadcrumb-eu.css',
  },
  shadow: false,
  scoped: true,
  assetsDirs: ['build']
})

export class EclBreadcrumb {
  @Element() el: HTMLElement;
  @Prop() styleClass: string = '';
  @Prop() variant: string = 'default';
  @Prop() theme: string = 'ec';
  @Prop() eclScript: boolean = false;
  @Prop() minItemsRight: number = 2;


  getClass(): string {
    return [
      `ecl-breadcrumb`,
      `ecl-breadcrumb--${this.variant}`,
      this.styleClass
    ].join(' ');
  }

  componentDidRender() {
    const items = this.el.querySelectorAll('.ecl-breadcrumb__segment');
    const itemsCount = items.length;
    const ellipsis = this.el.querySelector('.ecl-breadcrumb__segment--ellipsis');
    const startHidingIndex = Array.from(items).indexOf(ellipsis);
    const toBeHidden = Array.from(items).slice((startHidingIndex + 1), (itemsCount - this.minItemsRight));

    toBeHidden.forEach((item) => {
      item.setAttribute('data-ecl-breadcrumb-item', 'expandable');
    });

    this.el.querySelector('.ecl-breadcrumb__container').innerHTML = '';
    this.el.querySelector('.ecl-breadcrumb__container').append(...items);

    if (this.eclScript) {
      // Load the ECL vanilla js if not already present.
      const src = getAssetPath('./build/scripts/ecl-breadcrumb-vanilla.js');
      if (document.querySelector(`script[src="${src}"]`)) {
        document.querySelector(`script[src="${src}"]`).remove();
      }
      const script = document.createElement('script');
      script.src = src;
      script.onload = () => {
        const breadcrumb = new ECL.Breadcrumb(this.el.firstElementChild);
        breadcrumb.init();
      };

      document.body.appendChild(script);
    }
  }

  render() {
    return (
      <nav
        class={this.getClass()}
        data-ecl-breadcrumb
      >
        <ol class="ecl-breadcrumb__container">
          <slot></slot>
        </ol>
      </nav>
    )
  }
}

