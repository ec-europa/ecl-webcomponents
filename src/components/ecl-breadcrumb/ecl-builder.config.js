const path = require('path');
const pkg = require('./package.json');

const nodeModules = path.resolve(__dirname, './node_modules');
const outputFolder = './build';

// SCSS includePaths
const includePaths = [nodeModules];

const banner = `${pkg.name} - ${pkg.version} Built on ${new Date().toISOString()}`;

module.exports = {
  scripts: [
    {
      entry: path.resolve(nodeModules, '@ecl/vanilla-component-breadcrumb/breadcrumb.js'),
      dest: path.resolve(outputFolder, 'scripts/ecl-breadcrumb-vanilla.js'),
      options: {
        banner,
        moduleName: 'ECL',
        includePaths,
        sourceMap: false,
      },
    },
  ],
  styles: [
    {
      entry: path.resolve(__dirname, 'src/styles/ecl-breadcrumb-ec.scss'),
      dest: path.resolve(outputFolder, 'styles/ecl-breadcrumb-ec.css'),
      options: {
        banner,
        includePaths,
        sourceMap: false,
        minify: true,
      },
    },
    {
      entry: path.resolve(__dirname, 'src/styles/ecl-breadcrumb-eu.scss'),
      dest: path.resolve(outputFolder, 'styles/ecl-breadcrumb-eu.css'),
      options: {
        banner,
        includePaths,
        sourceMap: false,
        minify: true,
      },
    },
  ],
};
