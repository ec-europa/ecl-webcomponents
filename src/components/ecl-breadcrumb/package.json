{
  "name": "@ecl/web-components-breadcrumb",
  "author": "European Commission",
  "license": "EUPL-1.2",
  "version": "0.1.0",
  "description": "ECL Breadcrumb",
  "main": "ecl-breadcrumb.js",
  "private": true,
  "scripts": {
    "build:scripts": "cross-env ecl-builder scripts",
    "build:styles": "cross-env NODE_ENV=production ecl-builder styles",
    "build": "run-s clean build:*",
    "clean": "rimraf build",
    "dist": "cross-env NODE_ENV=production run-p clean build:*",
    "prepublish": "npm run build "
  },
  "devDependencies": {
    "@ecl/builder": "3.7.1",
    "@ecl/theme-ec": "3.7.1",
    "@ecl/theme-eu": "3.7.1",
    "@ecl/vanilla-component-breadcrumb": "3.7.1",
    "cross-env": "7.0.3"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/planctus/ecl-webcomponents.git"
  },
  "bugs": {
    "url": "https://github.com/planctus/ecl-webcomponents/issues"
  },
  "homepage": "https://github.com/planctus/ecl-webcomponents",
  "keywords": [
    "ecl",
    "europa-component-library",
    "design-system",
    "web-components"
  ]
}
