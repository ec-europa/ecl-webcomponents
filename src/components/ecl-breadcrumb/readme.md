# ecl-breadcrumb-item



<!-- Auto Generated Below -->


## Properties

| Property          | Attribute           | Description | Type      | Default     |
| ----------------- | ------------------- | ----------- | --------- | ----------- |
| `buttonAriaLabel` | `button-aria-label` |             | `string`  | `''`        |
| `currentPage`     | `current-page`      |             | `boolean` | `false`     |
| `ellipsis`        | `ellipsis`          |             | `boolean` | `false`     |
| `path`            | `path`              |             | `string`  | `''`        |
| `styleClass`      | `style-class`       |             | `string`  | `''`        |
| `theme`           | `theme`             |             | `string`  | `undefined` |


## Dependencies

### Used by

 - [ecl-breadcrumb](.)

### Depends on

- [ecl-link](../ecl-link)
- [ecl-button](../ecl-button)
- [ecl-icon](../ecl-icon)

### Graph
```mermaid
graph TD;
  ecl-breadcrumb-item --> ecl-link
  ecl-breadcrumb-item --> ecl-button
  ecl-breadcrumb-item --> ecl-icon
  ecl-link --> ecl-icon
  ecl-button --> ecl-indicator
  ecl-breadcrumb --> ecl-breadcrumb-item
  style ecl-breadcrumb-item fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
