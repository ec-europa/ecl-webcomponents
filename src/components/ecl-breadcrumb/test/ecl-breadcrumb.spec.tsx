import { newSpecPage } from '@stencil/core/testing';
import { EclBreadcrumb } from '../ecl-breadcrumb';
import { EclBreadcrumbItem } from '../ecl-breadcrumb-item';

describe('ecl-breadcrumb', () => {
  it('renders with ellipsis and current page', async () => {
    const page = await newSpecPage({
      components: [EclBreadcrumb, EclBreadcrumbItem],
      html: `<ecl-breadcrumb
        min-items-left="1"
        min-items-right="2"
        ecl-script="true"
        navigation-text="Breadcrumb"
        ellipsis-label="Expand breadcrumb"
      >
        <ecl-breadcrumb-item path="/example.html">
          Home
        </ecl-breadcrumb-item>
        <ecl-breadcrumb-item
          path="/example.html"
        >
          About the European Commission
        </ecl-breadcrumb-item>
        <ecl-breadcrumb-item
          path="/example.html"
        >
          Organisational structure
        </ecl-breadcrumb-item>
        <ecl-breadcrumb-item
          path="/example.html"
        >
          How the Commission is organised
        </ecl-breadcrumb-item>
        <ecl-breadcrumb-item
          current-page
        >
          News
        </ecl-breadcrumb-item>
      </ecl-breadcrumb>`,
    });

    expect(page.root).toMatchSnapshot();
  });
});
