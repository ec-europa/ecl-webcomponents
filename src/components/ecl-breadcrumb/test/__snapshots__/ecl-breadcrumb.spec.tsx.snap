// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ecl-breadcrumb renders with ellipsis and current page 1`] = `
<ecl-breadcrumb ecl-script="true" ellipsis-label="Expand breadcrumb" min-items-left="1" min-items-right="2" navigation-text="Breadcrumb">
  <!---->
  <nav aria-label="Breadcrumb" class="ecl-breadcrumb">
    <ol class="ecl-breadcrumb__container">
      <ecl-breadcrumb-item data-ecl-breadcrumb-item="static" path="/example.html">
        <!---->
        <li class="ecl-breadcrumb__segment sc-ecl-breadcrumb-ec">
          <ecl-link path="/example.html" style-class="ecl-link sc-ecl-link-ec sc-ecl-breadcrumb-ec ecl-link--standalone ecl-link--no-visited ecl-breadcrumb__link" variant="standalone">
            Home
          </ecl-link>
          <ecl-icon icon="corner-arrow" role="presentation" size="fluid" slot="icon-after" style-class="ecl-breadcrumb__icon sc-ecl-breadcrumb-ec" transform="rotate-90"></ecl-icon>
        </li>
      </ecl-breadcrumb-item>
      <ecl-breadcrumb-item button-aria-label="Expand breadcrumb" ellipsis="true">
        <!---->
        <li aria-hidden="true" class="ecl-breadcrumb__segment ecl-breadcrumb__segment--ellipsis sc-ecl-breadcrumb-ec" data-ecl-breadcrumb-ellipsis>
          <ecl-button aria-label="Expand breadcrumb" data-ecl-breadcrumb-ellipsis-button style-class="ecl-breadcrumb__ellipsis sc-ecl-breadcrumb-ec" variant="ghost">
            ...
          </ecl-button>
          <ecl-icon icon="corner-arrow" role="presentation" size="fluid" slot="icon-after" style-class="ecl-breadcrumb__icon sc-ecl-breadcrumb-ec" transform="rotate-90"></ecl-icon>
        </li>
      </ecl-breadcrumb-item>
      <ecl-breadcrumb-item aria-hidden="false" data-ecl-breadcrumb-item="expandable" path="/example.html">
        <!---->
        <li class="ecl-breadcrumb__segment sc-ecl-breadcrumb-ec">
          <ecl-link path="/example.html" style-class="ecl-link sc-ecl-link-ec sc-ecl-breadcrumb-ec ecl-link--standalone ecl-link--no-visited ecl-breadcrumb__link" variant="standalone">
            About the European Commission
          </ecl-link>
          <ecl-icon icon="corner-arrow" role="presentation" size="fluid" slot="icon-after" style-class="ecl-breadcrumb__icon sc-ecl-breadcrumb-ec" transform="rotate-90"></ecl-icon>
        </li>
      </ecl-breadcrumb-item>
      <ecl-breadcrumb-item aria-hidden="false" data-ecl-breadcrumb-item="expandable" path="/example.html">
        <!---->
        <li class="ecl-breadcrumb__segment sc-ecl-breadcrumb-ec">
          <ecl-link path="/example.html" style-class="ecl-link sc-ecl-link-ec sc-ecl-breadcrumb-ec ecl-link--standalone ecl-link--no-visited ecl-breadcrumb__link" variant="standalone">
            Organisational structure
          </ecl-link>
          <ecl-icon icon="corner-arrow" role="presentation" size="fluid" slot="icon-after" style-class="ecl-breadcrumb__icon sc-ecl-breadcrumb-ec" transform="rotate-90"></ecl-icon>
        </li>
      </ecl-breadcrumb-item>
      <ecl-breadcrumb-item data-ecl-breadcrumb-item="static" path="/example.html">
        <!---->
        <li class="ecl-breadcrumb__segment sc-ecl-breadcrumb-ec">
          <ecl-link path="/example.html" style-class="ecl-link sc-ecl-link-ec sc-ecl-breadcrumb-ec ecl-link--standalone ecl-link--no-visited ecl-breadcrumb__link" variant="standalone">
            How the Commission is organised
          </ecl-link>
          <ecl-icon icon="corner-arrow" role="presentation" size="fluid" slot="icon-after" style-class="ecl-breadcrumb__icon sc-ecl-breadcrumb-ec" transform="rotate-90"></ecl-icon>
        </li>
      </ecl-breadcrumb-item>
      <ecl-breadcrumb-item current-page="" data-ecl-breadcrumb-item="static">
        <!---->
        <li aria-current="page" class="ecl-breadcrumb__current-page ecl-breadcrumb__segment sc-ecl-breadcrumb-ec">
          News
        </li>
      </ecl-breadcrumb-item>
    </ol>
  </nav>
</ecl-breadcrumb>
`;
